<?xml version="1.0"?>
<sdf version="1.6">
  <model name="mycar">
    <include>
      <pose>0 0 0.05 0 0 0</pose>
      <uri>model://hokuyo_nogpu</uri>
    </include>
    <joint name="hokuyo_joint" type="fixed">
      <parent>chasis</parent>
      <child>hokuyo_nogpu::link</child>
    </joint>
    <link name="chasis">
      <inertial>
        <mass>1.0</mass>
        <inertia>
          <!-- inertias are tricky to compute -->
          <!-- http://gazebosim.org/tutorials?tut=inertia&cat=build_robot -->
          <ixx>0.1</ixx>
          <!-- for a box: ixx = 0.083 * mass * (y*y + z*z) -->
          <ixy>0</ixy>
          <!-- for a box: ixy = 0 -->
          <ixz>0</ixz>
          <!-- for a box: ixz = 0 -->
          <iyy>0.1</iyy>
          <!-- for a box: iyy = 0.083 * mass * (x*x + z*z) -->
          <iyz>0</iyz>
          <!-- for a box: iyz = 0 -->
          <izz>0.1</izz>
          <!-- for a box: izz = 0.083 * mass * (x*x + y*y) -->
        </inertia>
      </inertial>
      <collision name="collision">
        <geometry>
          <box>
            <size>0.5 0.2 0.1</size>
          </box>
        </geometry>
      </collision>
      <visual name="visual">
        <material>
          <ambient>0 0 0</ambient>
        </material>
        <geometry>
          <box>
            <size>0.5 0.2 0.1</size>
          </box>
        </geometry>
      </visual>
    </link>
    <link name="front_left_wheel">
      <pose>0.2 0.2 -0.1 1.57 0 0</pose>
      <inertial>
        <mass>0.1</mass>
        <inertia>
          <!-- inertias are tricky to compute -->
          <!-- http://gazebosim.org/tutorials?tut=inertia&cat=build_robot -->
          <ixx>0.01</ixx>
          <!-- for a box: ixx = 0.083 * mass * (y*y + z*z) -->
          <ixy>0</ixy>
          <!-- for a box: ixy = 0 -->
          <ixz>0</ixz>
          <!-- for a box: ixz = 0 -->
          <iyy>0.01</iyy>
          <!-- for a box: iyy = 0.083 * mass * (x*x + z*z) -->
          <iyz>0</iyz>
          <!-- for a box: iyz = 0 -->
          <izz>0.1</izz>
          <!-- for a box: izz = 0.083 * mass * (x*x + y*y) -->
        </inertia>
      </inertial>
      <collision name="collision">
        <geometry>
          <cylinder>
            <radius>0.05</radius>
            <length>0.04</length>
          </cylinder>
        </geometry>
      </collision>
      <visual name="visual">
        <material>
          <ambient>1 0 0</ambient>
        </material>
        <geometry>
          <cylinder>
            <radius>0.05</radius>
            <length>0.04</length>
          </cylinder>
        </geometry>
      </visual>
    </link>
    <joint name="front_left_axle" type="revolute2">
      <parent>chasis</parent>
      <child>front_left_wheel</child>
      <axis>
        <limit>
          <lower>-0.2</lower>
          <upper>0.2</upper>
        </limit>
        <xyz>0 0 1</xyz>
      </axis>
      <axis2>
        <xyz>0 1 0</xyz>
      </axis2>
    </joint>
    <link name="front_right_wheel">
      <pose>0.2 -0.2 -0.1 1.57 0 0</pose>
      <inertial>
        <mass>0.1</mass>
        <inertia>
          <!-- inertias are tricky to compute -->
          <!-- http://gazebosim.org/tutorials?tut=inertia&cat=build_robot -->
          <ixx>0.01</ixx>
          <!-- for a box: ixx = 0.083 * mass * (y*y + z*z) -->
          <ixy>0</ixy>
          <!-- for a box: ixy = 0 -->
          <ixz>0</ixz>
          <!-- for a box: ixz = 0 -->
          <iyy>0.01</iyy>
          <!-- for a box: iyy = 0.083 * mass * (x*x + z*z) -->
          <iyz>0</iyz>
          <!-- for a box: iyz = 0 -->
          <izz>0.1</izz>
          <!-- for a box: izz = 0.083 * mass * (x*x + y*y) -->
        </inertia>
      </inertial>
      <collision name="collision">
        <geometry>
          <cylinder>
            <radius>0.05</radius>
            <length>0.04</length>
          </cylinder>
        </geometry>
      </collision>
      <visual name="visual">
        <material>
          <ambient>1 0 0</ambient>
        </material>
        <geometry>
          <cylinder>
            <radius>0.05</radius>
            <length>0.04</length>
          </cylinder>
        </geometry>
      </visual>
    </link>
    <joint name="front_right_axle" type="revolute2">
      <parent>chasis</parent>
      <child>front_right_wheel</child>
      <axis>
        <limit>
          <lower>-0.2</lower>
          <upper>0.2</upper>
        </limit>
        <xyz>0 0 1</xyz>
      </axis>
      <axis2>
        <xyz>0 1 0</xyz>
      </axis2>
    </joint>
    <link name="rear_left_wheel">
      <pose>-0.2 0.2 -0.1 1.57 0 0</pose>
      <inertial>
        <mass>0.1</mass>
        <inertia>
          <!-- inertias are tricky to compute -->
          <!-- http://gazebosim.org/tutorials?tut=inertia&cat=build_robot -->
          <ixx>0.01</ixx>
          <!-- for a box: ixx = 0.083 * mass * (y*y + z*z) -->
          <ixy>0</ixy>
          <!-- for a box: ixy = 0 -->
          <ixz>0</ixz>
          <!-- for a box: ixz = 0 -->
          <iyy>0.01</iyy>
          <!-- for a box: iyy = 0.083 * mass * (x*x + z*z) -->
          <iyz>0</iyz>
          <!-- for a box: iyz = 0 -->
          <izz>0.1</izz>
          <!-- for a box: izz = 0.083 * mass * (x*x + y*y) -->
        </inertia>
      </inertial>
      <collision name="collision">
        <geometry>
          <cylinder>
            <radius>0.05</radius>
            <length>0.04</length>
          </cylinder>
        </geometry>
      </collision>
      <visual name="visual">
        <material>
          <ambient>0 0 1</ambient>
        </material>
        <geometry>
          <cylinder>
            <radius>0.05</radius>
            <length>0.04</length>
          </cylinder>
        </geometry>
      </visual>
    </link>
    <joint name="rear_left_axle" type="revolute">
      <parent>chasis</parent>
      <child>rear_left_wheel</child>
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>
    <link name="rear_right_wheel">
      <pose>-0.2 -0.2 -0.1 1.57 0 0</pose>
      <inertial>
        <mass>0.1</mass>
        <inertia>
          <!-- inertias are tricky to compute -->
          <!-- http://gazebosim.org/tutorials?tut=inertia&cat=build_robot -->
          <ixx>0.01</ixx>
          <!-- for a box: ixx = 0.083 * mass * (y*y + z*z) -->
          <ixy>0</ixy>
          <!-- for a box: ixy = 0 -->
          <ixz>0</ixz>
          <!-- for a box: ixz = 0 -->
          <iyy>0.01</iyy>
          <!-- for a box: iyy = 0.083 * mass * (x*x + z*z) -->
          <iyz>0</iyz>
          <!-- for a box: iyz = 0 -->
          <izz>0.1</izz>
          <!-- for a box: izz = 0.083 * mass * (x*x + y*y) -->
        </inertia>
      </inertial>
      <collision name="collision">
        <geometry>
          <cylinder>
            <radius>0.05</radius>
            <length>0.04</length>
          </cylinder>
        </geometry>
      </collision>
      <visual name="visual">
        <material>
          <ambient>0 0 1</ambient>
        </material>
        <geometry>
          <cylinder>
            <radius>0.05</radius>
            <length>0.04</length>
          </cylinder>
        </geometry>
      </visual>
    </link>
    <joint name="rear_right_axle" type="revolute">
      <parent>chasis</parent>
      <child>rear_right_wheel</child>
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>
    <plugin name="p3d" filename="libgazebo_ros_p3d.so">
      <bodyName>chasis</bodyName>
      <topicName>mycar/ground_truth</topicName>
      <frameName>world</frameName>
      <updateRate>100</updateRate>
    </plugin>
    <!--<plugin name="model_push" filename="libmodel_push.so"/>-->
    <plugin name="force_front_left" filename="libgazebo_ros_force.so">
      <alwaysOn>true</alwaysOn>
      <update>100</update>
      <updateRate>100.0</updateRate>
      <bodyName>front_left_wheel</bodyName>
      <topicName>mycar/force_front_left</topicName>
    </plugin>

    <plugin name="force_front_right" filename="libgazebo_ros_force.so">
      <alwaysOn>true</alwaysOn>
      <update>100</update>
      <updateRate>100.0</updateRate>
      <bodyName>front_right_wheel</bodyName>
      <topicName>mycar/force_front_right</topicName>
    </plugin>

    <plugin name="force_rear_left" filename="libgazebo_ros_force.so">
      <alwaysOn>true</alwaysOn>
      <update>100</update>
      <updateRate>100.0</updateRate>
      <bodyName>rear_left_wheel</bodyName>
      <topicName>mycar/force_rear_left</topicName>
    </plugin>

    <plugin name="force_rear_right" filename="libgazebo_ros_force.so">
      <alwaysOn>true</alwaysOn>
      <update>100</update>
      <updateRate>100.0</updateRate>
      <bodyName>front_right_wheel</bodyName>
      <topicName>mycar/force_rear_right</topicName>
    </plugin>

  </model>
</sdf>
<!-- vim: set et ft=xml fenc=utf-8 ff=unix sts=0 sw=2 ts=2 : -->
